ENH/PERF: enable column-wise reductions for EA-backed columns

jorisvandenbossche opened the pull request:
Currently, for reductions on a DataFrame, we convert the full DataFrame to a single "interleaved" array and then perform the operation. That's the default, but when `numeric_only=True` is specified, it is done block-wise.

Enabling column-wise reductions (or block-wise for EAs):

- Gives better performance in common cases / no need to create a new 2D array (which goes through object dtype for nullable ints)
- Ensures to use the reduction implementation of the EA itself, which can be more correct / more efficient than converting to an ndarray and using our generic nanops.

For illustration purposes, I added a `column_wise` keyword in this PR (not meant to keep this, just for testing), so we can compare a few cases:

```
In [9]: df_wide = pd.DataFrame(np.random.randint(1000, size=(1000,100))).astype("Int64").copy() 

In [15]: %timeit df_wide.mean()   
9.68 ms ± 100 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

In [16]: %timeit df_wide.mean(numeric_only=True)
10.1 ms ± 345 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

In [17]: %timeit df_wide.mean(column_wise=True)  
5.22 ms ± 29.7 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

In [18]: df_long = pd.DataFrame(np.random.randint(1000, size=(10000,10))).astype("Int64").copy()  

In [19]: %timeit df_long.mean()
7.77 ms ± 115 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

In [20]: %timeit df_long.mean(numeric_only=True)         
2.07 ms ± 31.3 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

In [21]: %timeit df_long.mean(column_wise=True) 
1.04 ms ± 4.54 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
```

So I experimented with two approaches:

- First by iterating through the columns and calling the `_reduce` of the underlying EA (this path gets taken by using the temporary keyword `column_wise=True`)
- Fixing the block-wise case for extension blocks (triggered by `numeric_only=True`) by changing that to also use `_reduce` of the EA (currently this was failing by calling nanops functions on the EA)

The first gives better performance (it is simpler in implementation by not involding the blocks), but requires some more new code (it uses less the existing machinery).

Ideally, for EA columns, we should always use their own reduction implementation (thus call `EA._reduce`), I think. So for both approaches, the question will be how to trigger this behaviour.


Closes #32651,  closes #34520

jorisvandenbossche committed:
ENH/PERF: enable column-wise reductions for EA-backed columns

jorisvandenbossche committed:
fix numeric_only for EAs

jorisvandenbossche commented:
cc @jbrockmendel

jbrockmendel commented on pandas/core/frame.py:
```diff
@@ -7898,6 +7915,13 @@ def _get_data(axis_matters):
+            df = self
+            if numeric_only is True:
+                df = _get_data(axis_matters=True)
+            return DataFrame._reduce_columns(df, op, name, skipna=skipna, **kwds)
```
is this different from `df._reduce_columns(op, name, skipna=skipna, **kwds)`?

jorisvandenbossche replied:
Ah, no, of course not, that's an oversight

jbrockmendel commented on pandas/core/frame.py:
```diff
@@ -7852,6 +7852,23 @@ def _count_level(self, level, axis=0, numeric_only=False):
+        skipna, **kwds : keywords to pass to the `op` function
+
+        """
+        column_wise = kwds.pop("column_wise", False)
```
any reason not to put this directly in the signature?

jorisvandenbossche replied:
See my top post comment, this is just for testing for now, so I didn't care much where to put it

jbrockmendel commented on pandas/core/frame.py:
```diff
@@ -7916,7 +7940,7 @@ def blk_func(values):
             # After possibly _get_data and transposing, we are now in the
             #  simple case where we can use BlockManager._reduce
-            res = df._data.reduce(blk_func)
+            res = df._data.reduce(blk_func, name, skipna, **kwds)
```
why do we need skipna and kwds again? they are already in blk_func

jorisvandenbossche replied:
`blk_func` is not used for EAs. But, can also put the `EA._reduce` call inside `blk_func`, then it doesn't need to be passed through

jorisvandenbossche committed:
fix _reduce_columns call

jorisvandenbossche commented on pandas/core/frame.py:
```diff
@@ -7994,6 +8018,47 @@ def blk_func(values):
+        for arr in self._iter_arrays():
+            if isinstance(arr, ExtensionArray):
+                # dispatch to ExtensionArray interface
+                val = arr._reduce(name, skipna=skipna, **kwds)
```
I would say it is up to the EA to provide a fallback, if they want. 

(assuming we call this only on numeric columns by default)

jbrockmendel commented on pandas/core/frame.py:
```diff
@@ -7994,6 +8018,47 @@ def blk_func(values):
+        result = []
+
+        for arr in self._iter_arrays():
+            if isinstance(arr, ExtensionArray):
```
could this go inside `blk_func`?  i think that would avoid the need for a bunch of edits.

jorisvandenbossche replied:
Is suppose so, yes, I just commented the same above

jorisvandenbossche replied:
done

jbrockmendel commented on pandas/core/internals/managers.py:
```diff
@@ -349,24 +349,27 @@ def _verify_integrity(self) -> None:
         for blk in self.blocks:
-            bres = func(blk.values, *args, **kwargs)
+            if isinstance(blk, ExtensionBlock):
+                bres = blk.values._reduce(name, skipna=skipna, **kwds)
```
For dt64 and td64 do we want to use the DTA/TDA reductions?

jbrockmendel commented:
Come to think of it, the place where this dispatch belongs may be in the relevant nanops functions

jorisvandenbossche commented:
> Come to think of it, the place where this dispatch belongs may be in the relevant nanops functions

Personally, I think the nanops should be about ops on numpy arrays, and not deal with extension arrays

jbrockmendel commented on pandas/core/internals/managers.py:
```diff
@@ -998,6 +1001,14 @@ def iget(self, i: int) -> "SingleBlockManager":
             fastpath=True,
         )
 
+    def iget_values(self, i: int):
```
`-> ArrayLike`

If we do implement this new method, let's use it in `iget`

jorisvandenbossche replied:
> If we do implement this new method, let's use it in iget

I did that first, but `iget` needs both block and values for the rest of its function, so using it there doesn't duplicate any code

jbrockmendel commented on pandas/core/frame.py:
```diff
@@ -7994,6 +8018,47 @@ def blk_func(values):
+
+        return self._constructor_sliced(result, index=self.columns)
+
+    def _iter_arrays(self):
```
if this is only used once, id rather inline it.  if we do add this method, it should probably go adjacent to [similar methods, not sure off the top of my head]

jorisvandenbossche replied:
Yeah, the location where to put it needs to be decided (if we keep it), just put it here because I was editing the code here.

I think it will be useful for more things in the future (anything that we want to optimize for column-wise), but we can also extract it at that moment that we would need it

jorisvandenbossche replied:
Inlined for now

jorisvandenbossche committed:
move EA._reduce call into blk_func

jorisvandenbossche committed:
reuse blk_func for column-wise, inline _iter_arrays

jbrockmendel commented on pandas/core/frame.py:
```diff
@@ -7898,6 +7915,19 @@ def _get_data(axis_matters):
             return data
 
+        def blk_func(values):
+            if isinstance(values, ExtensionArray):
```
with this inside blk_func, shouldn't the block-wise operation have the same performance bump as the column-wise?

jorisvandenbossche replied:
> with this inside blk_func, shouldn't the block-wise operation have the same performance bump as the column-wise?

It didn't actually change anything performance wise (it's the same function being called as before). 
The reason that both paths have different performance, is because the re-assembling of the results into a Series is more expensive for the block-wise compared to column-wise. 

(it's possible that the block-wise way could be optimized to get rid of this difference though. The main thing is that the block results are not in order)

jbrockmendel replied:
> (it's possible that the block-wise way could be optimized to get rid of this difference though. The main thing is that the block results are not in order)

This would be really nice.

jorisvandenbossche commented:
So the bigger question is: how can we get to use this by default (at least for EAs). Some ideas:

- The `column_wise` route could for example be triggered by checking if all columns are ExtensionBlock columns (we already have a `BlockManager.any_extension_types`, could also have an `all_extension_types`). 
  This way, this would only be triggered in rare cases though (certainly as long we don't yet have a float extension type). We could also trigger it with `any_extension_types`: for the new nullable dtype that would be possible (since those are new, they can introduces changes in behaviour), but not for the older extension dtypes (categorical, datetimelikes)

- Currently, with `numeric_only is None`, we first try on the full values (`DataFrame.values`), and then fall back to `frame_apply` with `ignore_failures=True`. 
  I *suppose* we could also do this block-wise, just there ignore failures in case of `numeric_only is None`, and in such a case skip those blocks to assemble the output (but will need to check what all fails by doing this ..). 
  If that works, we could get rid of the full `numeric_only is None`-block in `_reduce`.

jbrockmendel commented:
> Currently, with numeric_only is None, we first try on the full values (DataFrame.values), and then fall back to frame_apply with ignore_failures=True.

I've got a branch that identifies the cases where frame_apply is used and does that before the .values call.  Trying to 1) avoid a .values call and 2) de-nest _reduce.  I'll move that branch up the priority-list.

jbrockmendel commented:
> Personally, I prefer the nanops to be about ops on numpy arrays, and not deal with extension arrays

Depends on the public-ness of those functions.  ATM their docstrings have examples with Series, not sure if we have other docs or tests with those.

jorisvandenbossche commented:
> I've got a branch that identifies the cases where frame_apply is used and does that before the .values call. Trying to 1) avoid a .values call and 2) de-nest _reduce. I'll move that branch up the priority-list.

Note that my suggestion was to eliminate this part entirely (by using block-wise for all). So let's ensure we don't do duplicate / conflicting work. Does the branch already have something? (can you maybe push it to your fork?)

> Depends on the public-ness of those functions. ATM their docstrings have examples with Series, not sure if we have other docs or tests with those.

nanops are not public (regardless of their docstrings). What I meanly meant is that (in my head) they are meant to work on numpy arrays (whether extracted from a Series first or not)

jbrockmendel commented:
https://github.com/jbrockmendel/pandas/tree/cln-reduce

> Note that my suggestion was to eliminate this part entirely (by using block-wise for all).

This would require making `BlockManager._reduce` handle `numeric_only=None` similar to how `frame_apply` handles `allow_failures=True`, right?  I've been reticent to do that (because I dont like the ignoring-exceptions pattern, xref #28900), but it may end up being the best option (off the top of my head, this would make it easy to solve #28773).

jorisvandenbossche committed:
temp

jorisvandenbossche commented:
Thanks!

> because I dont like the ignoring-exceptions pattern,

Yes, but as long as we keep the `numeric_only=None` behaviour, that's the whole point of it .. (to ignore exceptions). So it would still be nice to try to unify the two code paths a bit more than what we have now. 
Will try to take a look at it next week.

jbrockmendel commented:
hmm block-wise and column-wise with ignore_failures wont necessarily be equivalent for object-dtype

jorisvandenbossche commented:
> hmm block-wise and column-wise with ignore_failures wont necessarily be equivalent for object-dtype

Ah, that's a good point. So for ObjectBlock, we would still need to do it column-wise, if we want to get rid of the fallback in general. Will take a further look one of the coming days if that looks feasable

jorisvandenbossche committed:
Merge remote-tracking branch 'upstream/master' into EA-column-wise-reductions

jorisvandenbossche committed:
first attempts of going block-wise with numeric_only=None

pep8speaks commented:
Hello @jorisvandenbossche! Thanks for updating this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:





* In the file [`pandas/core/internals/managers.py`](https://github.com/pandas-dev/pandas/blob/b6a94b582c36be6ad49b8ad4030a2b272ec49688/pandas/core/internals/managers.py):

> [Line 370:89](https://github.com/pandas-dev/pandas/blob/b6a94b582c36be6ad49b8ad4030a2b272ec49688/pandas/core/internals/managers.py#L370): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (107 > 88 characters)

jorisvandenbossche commented:
@jbrockmendel did some further tests with trying to always use the block-wise path. Tests are far from passing, but I already included some special cases (like object block still needing to fall back to column-wise)

This will make the the block-wise path of course a *lot* more complex (given all those special cases, and with try/except blocks ...). So would like to get some feedback on this possible path forward before I put more time in it fixing all corner cases. 
I *think* it will make some special cases more explicit in the code, though, and the original goal was to get use the EA-native reduce by default, which needs either column or block wise. So some extra complexity will be inevitable for achieving that goal, I think (when preserving all special behaviours of `numeric_only=None`).

jorisvandenbossche commented on pandas/core/frame.py:
```diff
@@ -7971,35 +7988,67 @@ def _get_data(axis_matters):
+            if len(res) < len(df.columns):
+                out.index = df.columns[np.sort(list(res.keys()))]
+            else:
+                out.index = df.columns
```
This construction of `out` gets a bit more complicated, as it is now not guaranteed any more that all columns are present in the `res`.

jorisvandenbossche commented on pandas/core/internals/managers.py:
```diff
@@ -350,25 +351,62 @@ def _verify_integrity(self) -> None:
         for blk in self.blocks:
-            bres = func(blk.values, *args, **kwargs)
+            placement = blk.mgr_locs.as_array
+            if isinstance(blk, CategoricalBlock):
```
(we could also have a `Block.reduce` that can hold those Block-specific logic), to keep this BlockManager.reduce more straightforward)

jorisvandenbossche commented on pandas/core/internals/managers.py:
```diff
@@ -350,25 +351,62 @@ def _verify_integrity(self) -> None:
+                    # Categorical, so fallback to operating on dense array
+                    # eg pandas/tests/frame/test_analytics.py::TestDataFrameAnalytics::test_any_all_np_func
+                    bres = func(np.asarray(blk.values).reshape(1, len(blk.values)))
+            elif isinstance(blk, ObjectBlock):
```
This long block is to handle object dtype column-wise when it fails on all values (since it object can be mixed). Could also go directly to column-wise always avoiding the first try/except.

jorisvandenbossche commented:
Alternative of course could be to not try to put everything in the block-wise case, but just go with column-wise instead of block-wise. But that would mean a perf degradation for 2D blocks. Unless we would do the column-wise for `numeric_only=None` and the block-wise for `numeric_only=True/False`, but that would give some duplication.

jbrockmendel commented on pandas/core/internals/managers.py:
```diff
@@ -350,25 +351,62 @@ def _verify_integrity(self) -> None:
+                except TypeError:
+                    # object dtype can have different type of objects in
+                    # different columns, so for this specific case we need
+                    # to fall back to apply the function column-wise
```
can we use `_split_and_operate` here?

jbrockmendel commented:
> Unless we would do the column-wise for numeric_only=None and the block-wise for numeric_only=True/False, but that would give some duplication.

I think this makes a lot of sense.  ATM we only do block-wise for `numeric_only is not None and axis in [0, 1]`.  It sounds like there is room to optimize the re-comining-results step, but the code for this is pretty clean.

We also already have frame_apply that supports both column-wise and allow_failures.  ATM we only use frame_apply as a fall-back.  Can we pick up the performance gains you have in mind by using it more aggressively?

Going back to the OP:

> Ideally, for EA columns, we should always use their own reduction implementation (thus call EA._reduce), I think. So for both approaches, the question will be how to trigger this behaviour.

Can we do this independent of the rest of this PR (and the difficult decisions it entails) with a small edit to `blk_func`?

jorisvandenbossche commented:
> Can we do this independent of the rest of this PR (and the difficult decisions it entails) with a small edit to blk_func?

No, because that is (in the current state of master) only used for `numeric_only=True/False`, and I want EA._reduce be used by default. 

(the small edit to blk_func is still useful for the  `numeric_only=True/False`, for sure)

jorisvandenbossche commented:
> We also already have frame_apply that supports both column-wise and allow_failures. ATM we only use frame_apply as a fall-back. Can we pick up the performance gains you have in mind by using it more aggressively?

Yeah, that might be a good idea. Will need to take a look to see if that is possible.

jbrockmendel commented:
> (the small edit to blk_func is still useful for the numeric_only=True/False, for sure)

Yah, thats what i was suggesting

jorisvandenbossche committed:
Merge remote-tracking branch 'upstream/master' into EA-column-wise-reductions

jorisvandenbossche committed:
TEMP

jbrockmendel commented:
@jorisvandenbossche is this still active?

jorisvandenbossche commented:
Yes, just need to find the time to investigate the `frame_apply` option (I already looked at it briefly, and my first impression is that it is doing way too much things that are all not needed if you know you are doing a simple reduction, but need to check the impact of this)

jorisvandenbossche committed:
Merge remote-tracking branch 'upstream/master' into EA-column-wise-reductions

jbrockmendel commented:
gentle encouragement; i think this could make a significant improvement

jorisvandenbossche committed:
use iter_column_arrays

jorisvandenbossche committed:
Merge remote-tracking branch 'upstream/master' into EA-column-wise-reductions

jbrockmendel commented on pandas/core/frame.py:
```diff
@@ -8604,6 +8604,7 @@ def blk_func(values):
 
             # breakpoint()
```
leftover?

jorisvandenbossche commented:
I would hold off reviewing for a moment, it is in a totally messy state, I am cleaning it up right now (I only rebased up to now)

jorisvandenbossche committed:
intermediate clean-up: remove BM.reduce changes + do column-wise for axis=0

jorisvandenbossche committed:
fixup

jorisvandenbossche committed:
fix dtype of empty result

jorisvandenbossche committed:
Merge remote-tracking branch 'upstream/master' into EA-column-wise-reductions

jorisvandenbossche committed:
clean-up

jorisvandenbossche committed:
whitespace

jorisvandenbossche commented:
Updated / cleaned up this branch. I now implemented column (array) wise for `axis=0` when the block-wise approach is not used. 
To limit potentially unforeseen backwards incompatible changes, we could also only take this route if there is at least one extension array column (or more restrictive: any extension array column with nullable dtype).

Remaining failures are all dubious cases (for which the solution might also be to fix the test):

- Any / all on timedelta/datetime dtypes -> https://github.com/pandas-dev/pandas/issues/34479
- Reduction on empty object-dtype DataFrame, currently returns float for DataFrame but integer for Series:

    ```
    In [6]: df_empty = pd.DataFrame(columns=['a', 'b', 'c']) 

    In [8]: df_empty.dtypes 
    Out[8]: 
    a    object
    b    object
    c    object
    dtype: object

    In [9]: df_empty.sum()  
    Out[9]: 
    a    0.0
    b    0.0
    c    0.0
    dtype: float64

    In [10]: s_empty = pd.Series(dtype=object)  

    In [11]: s_empty.sum() 
    Out[11]: 0

    In [12]: type(_)   
    Out[12]: int
    ```
  while on this PR the DataFrame behaviour follows Series to return integer series.

jorisvandenbossche commented on pandas/core/frame.py:
```diff
@@ -8559,6 +8580,54 @@ def blk_func(values):
+
         if not self._is_homogeneous_type:
             # try to avoid self.values call
 
```
This block (using `frame_apply`) could be removed now, since this is only called when `axis=0`, which is now already handled above.

We *might* want to limit the new code above to `if axis == 0 and self._mgr.any_extension_types:`, and then keep the block here for backwards compatibility (or eg only remove when doing a major version bump pandas 2.0).

jbrockmendel commented:
> Remaining failures are all dubious cases (for which the solution might also be to fix the test):

Agreed those cases look sketchy.  Especially the one with DataFrame not matching Series.

> We might want to limit the new code above to if axis == 0 and self._mgr.any_extension_types:, and then keep the block here for backwards compatibility

Did you try the previously-discussed approach of going through frame_apply to operate blockwise?

jorisvandenbossche commented:
> Did you try the previously-discussed approach of going through frame_apply to operate blockwise?

Yes, I did, but forgot to add a note about this. I am planning to open some issues for this, but the main reason I didn't take that route of `frame_apply` now: it's rather complex code (it's more general / doing a lot more than just the reductions I need here), it's also not yet working for ExtensionArrays (which is of course something that needs to be fixed in itself), and those two things together mean that it would be more involved to properly fix `frame_apply` to serve the purpose for `DataFrame._reduce`.  In addition, `frame_apply` also doesn't give the same performance increase as the current code in the PR (in part because it is doing a lot more).

We should certainly improve `frame_apply` as well (and at that point, the code I am adding here might be movable to `frame_apply` almost in its entirety), but since properly supporting EAs there is a larger endeavor, I went for now to fix it only here for `DataFrame._reduce`, as it would be nice to at least already start using the EA reductions in the DataFrame reductions.

jbrockmendel commented on pandas/core/frame.py:
```diff
@@ -8531,7 +8547,12 @@ def _get_data(axis_matters):
-        if numeric_only is not None and axis in [0, 1]:
+        # special case for block-wise
+        if (
+            not self._mgr.any_extension_types
```
why are we excluding any_extension_types here?  `blk_func` looks like it should handle EAs appropriately

jorisvandenbossche replied:
It indeed handles EA correctly (that fix was split off from this PR a while ago), but the reason I put this check here: the motivation for this branch (do block-wise when possible because it is faster) is not the case for EAs, as block-wise is actually slower. And secondly, that also ensures there is only a single code path when having EAs. 

Now, we could also only check for the nullable extension type (nullable int, float, bool etc, so it only checks for the new EAs), and not for all extension types in general (which then also includes things like datetimetz)

jorisvandenbossche replied:
@jbrockmendel I reverted this to keep the original check (so the same cases will take the block-wise code path as before this PR)

jbrockmendel replied:
thanks, ill take a fresh look

jbrockmendel commented on pandas/core/frame.py:
```diff
@@ -8559,6 +8580,54 @@ def blk_func(values):
+            if isinstance(values, ExtensionArray):
+                return values._reduce(name, skipna=skipna, **kwds)
+            else:
+                return op(values, skipna=skipna, **kwds)
```
why not pass axis here?  that looks like the only difference between this and blk_func

jorisvandenbossche replied:
Yes, blocks are 2D and thus the `blk_func` needs an axis, while this is working on the "column arrays", which are always 1D

jbrockmendel replied:
would passing axis=0 be harmful?  if not, can we de-duplicate these?

jorisvandenbossche replied:
> would passing axis=0 be harmful? if not, can we de-duplicate these?

For blocks, the axis is flipped, since the array is oriented differently. So if we share it, it would need some logic for which value of `axis` to pass. At which point the benefit of sharing might be lost.

jbrockmendel commented on pandas/core/frame.py:
```diff
@@ -8559,6 +8580,54 @@ def blk_func(values):
+                try:
+                    return _reduce_columns(df, array_func)
+                except TypeError:
+                    # if column-wise fails and numeric_only was None, we try
```
is the try/except here needed?  why not go straight to the except?

jorisvandenbossche replied:
Ah, good point. I think I originally did this in the idea that "this is faster for the cases when no exceptions would be raised". But I suppose the cost of also keeping track of the `indices` (which is the main difference with `_reduce_columns`) isn't very big. Will check.

In principle also the `if numeric_only is True:` one could be combined, if we move the `numeric_only` check inside the try/except in the inner loop

jbrockmendel replied:
> In principle also the if numeric_only is True: one could be combined, if we move the numeric_only check inside the try/except in the inner loop

Have to admit I'm having trouble following all the cases.  Have you already split off pieces that can be addressed independently? (e.g. you mentioned td64/dt64 .all() and some empty-frame behavior)

jorisvandenbossche replied:
Removed the try/except here, directly use the code that was now in the except clause

jbrockmendel commented on pandas/core/frame.py:
```diff
@@ -8559,6 +8580,54 @@ def blk_func(values):
+                    for i, arr in enumerate(df._iter_column_arrays()):
+                        try:
+                            res = array_func(arr)
+                        except Exception:
```
can this be any tighter?

jorisvandenbossche replied:
We use the same Exception in `frame_apply`:

https://github.com/pandas-dev/pandas/blob/69cc851a2ad077529c1a1fd4a3484a24e6c32b50/pandas/core/apply.py#L336-L344

Now, in apply this can be any user provided function, so needs to be general. Here, we know we are calling `EA._reduce` or one of the nanops ops, so I *suppose* it can be tighter. I assume eg nanops raise TypeError if it is a dtype they don't support?

jbrockmendel replied:
Makes sense, thanks.

jbrockmendel commented on pandas/core/internals/managers.py:
```diff
@@ -327,24 +327,24 @@ def _verify_integrity(self) -> None:
                 f"tot_items: {tot_items}"
             )
 
-    def reduce(self, func, *args, **kwargs):
```
can the edits in this file be done independent of the rest of the PR?  (also generic.py i think)

jbrockmendel commented:
AFAICT there are two perf-based motivations: 1) avoiding realizing `frame.values` (mentioned in the "interleaving" part of the OP) and 2) overhead associated with constructing a Series from block-wise op.  Do you have a clear picture of the relative importance of these?

TomAugspurger commented:
Does this close https://github.com/pandas-dev/pandas/issues/34520?

jorisvandenbossche commented:
Yes, but might be good to add a test for it (will do that here)

jorisvandenbossche commented:
> AFAICT there are two perf-based motivations: 1) avoiding realizing frame.values (mentioned in the "interleaving" part of the OP) and 2) overhead associated with constructing a Series from block-wise op. Do you have a clear picture of the relative importance of these?

Regarding 1), avoiding working on `frame.values` is not only for performance, but also for ensuring we use the actual EA._reduce (which might be faster, but also simply more correct). 

Regarding 2), the overhead from constructing a Series from block-wise ops is real but also not huge (I don't remember the actual number, but blockwise is certainly already a significant improvement compared to master performance wise), but the main reason that I gave up on handling everything inside BlockManager._reduce (the attempt is still somewhere in the git history of this branch) was that it became quite complex. One of the reasons for this is that for ObjectBlock, we then need to do it column-wise anyway there, because different columns of an ObjectBlock might hold different types and have differently. 
In this commit I removed it again:  https://github.com/pandas-dev/pandas/pull/32867/commits/7088cfcbbf37cff20fe06432dd5062bcabfe1f09 (see the diff in managers.py)

Of course, now is the complexity (to handle potential failures etc) inside frame.py, and since it needs to live somewhere, it can certainly also be somewhere else.

jbrockmendel commented:
> which might be faster, but also simply more correct

+1 on "more correct" being the priority.

> the overhead from constructing a Series from block-wise ops is real but also not huge

This seems to contradict a [previous comment](https://github.com/pandas-dev/pandas/pull/32867#discussion_r436390017).  If the `if numeric_only is not None and axis in [0, 1]:` case is correct as-is and there is not a compelling perf reason to change it, I would prefer to keep it as is, as it significantly narrows down the set of cases to be considered in the rest of the method.

> Of course, now is the complexity (to handle potential failures etc) inside frame.py, and since it needs to live somewhere, it can certainly also be somewhere else.

longer-term we should try to re-use the existing machinery in `frame_apply` (i think `apply_series_generator` with `allow_failures=True`), but if that is burdensome short term, I'll take your word for it.

Two other medium-term options are #34714 and #28900.

jorisvandenbossche committed:
Merge remote-tracking branch 'upstream/master' into EA-column-wise-reductions

jreback commented on pandas/core/frame.py:
```diff
@@ -8481,6 +8481,22 @@ def _count_level(self, level, axis=0, numeric_only=False):
     ):
+        """
+        Reduce DataFrame over axis with given operation.
+
```
typing args a +

jreback reviewed (changes requested):
this is massivley increasing the complexity here. Please find a way to do better.

jorisvandenbossche committed:
Merge remote-tracking branch 'upstream/master' into EA-column-wise-reductions

jorisvandenbossche committed:
add test case for GH34520, copied from GH35112

Co-authored-by: Simon Hawkins <simonjayhawkins@gmail.com>

jorisvandenbossche committed:
add test to ensure EA op is used for integer array

jorisvandenbossche committed:
remove try except

jorisvandenbossche committed:
remove unused code

jorisvandenbossche committed:
add test for GH32651, copied from GH34210

Co-authored-by: Simon Hawkins <simonjayhawkins@gmail.com>

jorisvandenbossche committed:
remove check for EAs for block-wise path

jorisvandenbossche commented:
> this is massivley increasing the complexity here. Please find a way to do better.

Do you have something more concrete as feedback? I don't find this particularly complex. It adds some more code, for sure, to ensure we perform the reductions properly column-wise with the correct operation (which fixes bugs).

Note that there was still code that could be removed (so I was actually replacing something, not only adding). I removed that now.

jbrockmendel commented:
> Do you have something more concrete as feedback? 

A suggestion for the short-term, (i.e. to address #35112) is to change the inline-defined `f` to something like:

```
    def blk_func(values):
            if is_extension_array_dtype(values.dtype):
                return extract_array(values)._reduce(name, skipna=skipna, **kwds)
            else:
                return op(values, axis=axis, skipna=skipna, **kwds)
```

in a dedicated PR.  I think that would address a subset of what this PR is doing and is orthogonal to the rest of this.

jreback commented:
this needs to wait for 1.2

jreback commented:
moving to 1.2

jorisvandenbossche commented:
I am fine with that, if we then include https://github.com/pandas-dev/pandas/pull/35254 instead

jbrockmendel commented:
@jorisvandenbossche I think #36076 has a bearing on this, thoughts?

Dr-Irv commented:
> * Reduction on empty object-dtype DataFrame, currently returns float for DataFrame but integer for Series:
>...
>   while on this PR the DataFrame behaviour follows Series to return integer series.

But in current behavior, a similar thing happens when the `DataFrame` has `object` dtype, but consists of int values and is not empty.  Not sure what your PR does in this case.
```python
>>> ddf=pd.DataFrame([[1,2,3]],columns=['a','b','c'],dtype=object)
>>> ddf
   a  b  c
0  1  2  3
>>> ddf.sum()
a    1.0
b    2.0
c    3.0
dtype: float64
>>> ddf.dtypes
a    object
b    object
c    object
dtype: object
>>> ddf['a'].sum()
1
>>> type(_)
<class 'int'>
>>>
```

jorisvandenbossche commented:
@Dr-Irv ah, thanks, that's something we should then also test. And another thing to look into ;)

jbrockmendel commented:
@jorisvandenbossche i think this is closable, as it really isnt actionable until we get the #36076-and-similar inconsistencies fixed, and at that point we'll be able to go all blockwise (which for ArrayManager will be columnwise anyway)

jreback commented:
looks ok, but not for 1.2

jbrockmendel commented:
@jorisvandenbossche does this still have a perf impact given that we don't go through .values anymore for axis=0?  Or is the perf impact all in re-assembling blockwise results?

We now always use EA._reduce I think, so that part of the motivation should no longer be relevant.

As @jreback referred to in his [previous](https://github.com/pandas-dev/pandas/pull/32867#pullrequestreview-446386754) comment, we've gone to a lot of trouble to simplify DataFrame._reduce, i'm wary of adding more re-complexifying it.

jreback commented:
@jorisvandenbossche pls update or close

jbrockmendel commented:
is this still relevant?

simonjayhawkins commented:
@jorisvandenbossche closing as stale. reopen when ready.
